upgrade/src/lib/utils.ts

```diff
@@ -0,0 +1,156 @@
+import { Octokit } from "@octokit/rest";
+import { clsx, type ClassValue } from "clsx";
+import { twMerge } from "tailwind-merge";
+import { z } from "zod";
+
+import { env } from "~/env.mjs";
+
+export const cn = (...inputs: ClassValue[]) => {
+  return twMerge(clsx(inputs));
+};
+
+export type VersionsGroupedByMajor = Array<{
+  major: string;
+  versions: string[];
+}>;
+
+export const getT3Versions = async () => {
+  const response = await fetch(
+    "https://api.github.com/repos/t3-oss/create-t3-app/releases?per_page=100"
+  );
+
+  const responseSchema = z.array(z.object({ tag_name: z.string() }));
+  const parsed = responseSchema.safeParse(await response.json());
+
+  if (!parsed.success) {
+    return [];
+  }
+
+  return parsed.data
+    .map((release) => release.tag_name.split("@")[1] ?? "")
+    .filter((v) => v !== "");
+};
+
+export const getT3VersionsGroupedByMajor = async () => {
+  const actualVersions = await getT3Versions();
+
+  const versionsGroupedByMajor: VersionsGroupedByMajor = [];
+
+  actualVersions.forEach((version) => {
+    const [major] = version.split(".");
+    if (!major) return;
+
+    const majorGroup = versionsGroupedByMajor.find(
+      (group) => group.major === major
+    );
+
+    if (majorGroup) {
+      majorGroup.versions.push(version);
+    } else {
+      versionsGroupedByMajor.push({
+        major,
+        versions: [version],
+      });
+    }
+  });
+
+  return versionsGroupedByMajor.sort(
+    (a, b) => Number(b.major) - Number(a.major)
+  );
+};
+
+export interface Features {
+  nextAuth?: boolean;
+  prisma?: boolean;
+  trpc?: boolean;
+  tailwind?: boolean;
+}
+
+export const getFeaturesString = (features: Features) => {
+  return Object.entries(features)
+    .filter(([, value]) => value)
+    .map(([key]) => key)
+    .join("-");
+};
+
+export const getFeatureUrl = (feature: string) => {
+  if (feature === "nextAuth") {
+    return "https://next-auth.js.org/";
+  } else if (feature === "prisma") {
+    return "https://www.prisma.io/";
+  } else if (feature === "trpc") {
+    return "https://trpc.io/";
+  } else if (feature === "tailwind") {
+    return "https://tailwindcss.com/";
+  }
+};
+
+type VersionsRegex = {
+  currentVersion: string;
+  upgradeVersion: string;
+};
```

Comment:
i recognize some of this code from the diff repo, forgot to remove somthing or is it needed in both?

Reply:
It's needed in both. I removed the unnecessary methods already